### DISCUSSION TITLE
Union type field in an un-keyable entity gets wrong typename from the union

### DISCUSSION THREAD

**Discussion by gjorgi-krenkov** [urql-graphql/urql#3272] [Q&A]
So I have a fragment in this kind of style:

```
fragment fullOption on Option {
  selectId
  type
  options {
    ... on NumberArray {
      values
    }
    ... on CustomNumberArray {
      id
      name
      values
    }
  }
}
```
All of the `Option`, `NumberArray` and `CustomNumberArray` are un-keyable entities(considering the `CustomNumberArray` as keyable didn't fix the issue).

Invalidating the `Option` field is obviously overcoming this issue in the context of Mutations, but I have this problem occurring with the Queries where even though the API data has the correct typenames for the respective options, the URQL Cache client messes that data and there is an `Option` that is `NumberArray` but it later gets `CustomNumberArray` typename.

What's additionally weird isn't reproducible in a different environment.

Anyone has met with something similar?

**Comment by kitten** [comment]
I'm not fully sure I understand all of what you're saying and what issue this causes, however, importantly `values` here is present on `NumberArray` _and_ on `CustomNumberArray`. What this means is that during a query Graphcache will find that `NumberArray` will also match when the type is actually a `CustomNumberArray`. That's because of heuristic fragment matching.

Currently, the only way to fix this is to either rename the fields so they don't conflict, or to add schema information to Graphcache. I was considering adding some kind of alternative for unions at least, but the gist of the issue here is that, structurally, the smaller of the fragments will heuristically match.

**Reply by gjorgi-krenkov** [reply]
Hi @kitten ,

The issue is the other way around, the type is `NumberArray`, but Graphcache considers it as a `CustomNumberArray`, and yes we do have codegen instrospected schema provided to Graphcache.

Need to take a look though whether renaming might help in our scenario. Also one crucial thing is that `NumberArray` doesn't contain neither of `id` and `name` so it doesn't make sense to match that data as a `CustomNumberArray`.

Also in the devtools, all of the `NumberArray` elements show up with an `id` and `name` value of undefined.
